File: java/src/main/java/bool/EvalVisitor.java
package bool;

import java.util.List;

public class EvalVisitor implements Visitor<Boolean, List<String>> {
    @Override
    public Boolean visitVar(Var var, List<String> vars) {
        return vars.contains(var.name);
    }

    @Override
    public Boolean visitAnd(And and, List<String> vars) {
        return and.left.accept(this, vars) && and.right.accept(this, vars);
    }

    @Override
    public Boolean visitOr(Or or, List<String> vars) {
        return or.left.accept(this, vars) || or.right.accept(this, vars);
    }

    @Override
    public Boolean visitNot(Not not, List<String> vars) {
        return !not.node.accept(this, vars);
    }
}

File: ruby/lib/bool/eval_visitor.rb
module Bool
  class EvalVisitor
    if RUBY_PLATFORM =~ /java/
      require 'libbool'

      def self.new
        Java::Bool::EvalVisitor.new
      end

    else
      def visit_var(var_node, vars)
        !!vars.index(var_node.name)
      end

      def visit_and(and_node, vars)
        and_node.left.describe_to(self, vars) && and_node.right.describe_to(self, vars)
      end

      def visit_or(and_node, vars)
        and_node.left.describe_to(self, vars) || and_node.right.describe_to(self, vars)
      end

      def visit_not(not_node, vars)
        !not_node.other.describe_to(self, vars)
      end
    end
  end
end

